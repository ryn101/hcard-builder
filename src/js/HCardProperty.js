/**
 * Configured types of hCard properties used as a lookup for value safety
 *
 * @type {{
 *    country: string,
 *    streetAddress: string,
 *    postalCode: string,
 *    locality: string,
 *    telephone: string,
 *    region: string,
 *    email: string
 *  }}
 */
const hCardPropertyType = {
    email: 'email',
    telephone: 'tel',
    streetAddress: 'street-address',
    locality: 'locality',
    region: 'region',
    postalCode: 'postal-code',
    country: 'country',
};

/**
 * Container for a HCard specific property
 */
class HCardProperty {
    /**
     * Constructor
     *
     * @param name The property name/type
     * @param value The value of the property
     */
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
}

export { hCardPropertyType as HCardPropertyType };
export default HCardProperty;
