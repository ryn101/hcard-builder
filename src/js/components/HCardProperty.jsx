import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Wraps up the logic needed to display a hCard property. The name of the property is added
 * as a class to the component and the title is set to the specified value.
 *
 * Additionally supports property nesting, and abbreviation declaration for use with the abbr tag.
 * If the abbreviation tag is provided then the as property is disregarded.
 * If child elements are provided then the value and abbreviation properties are disregarded.
 *
 * When requiring a HTML tag other than a span, the as property can be specified.
 */
function HCardProperty(props) {
    const {
        name,
        value,
        abbreviation,
        as,
        className,
        children,
    } = props;

    const ElementTag = as;
    let element;

    if (children || !abbreviation) {
        // If children are specified or no abbreviation value is specified then use the element tag
        element = (
            <ElementTag title={value} className={classNames(name, className)}>
                { children || value }
            </ElementTag>
        );
    } else {
        // If an abbreviation is specified then create an abbr tag instead
        element = (<abbr title={value} className={name}>{abbreviation}</abbr>);
    }

    return element;
}

HCardProperty.propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    abbreviation: PropTypes.string,
    as: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.oneOfType([
            PropTypes.element,
            PropTypes.string,
        ])),
    ]),
};

HCardProperty.defaultProps = {
    value: '',
    as: 'span',
    className: '',
    children: null,
};

export default HCardProperty;
