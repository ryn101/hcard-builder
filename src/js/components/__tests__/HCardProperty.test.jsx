import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardProperty from '../HCardProperty';

configure({ adapter: new Adapter() });

it('renders the hCard property without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<HCardProperty name="" />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once the hcard property has rendered', () => {
    let hCardPropertyNode;
    const propertyName = 'property-name';
    const propertyValue = 'Property Value';

    beforeEach(() => {
        hCardPropertyNode = shallow(
            <HCardProperty name={propertyName} value={propertyValue} />,
        );
    });

    it('renders with state correctly', () => {
        expect(hCardPropertyNode).toMatchSnapshot();
    });

    it('renders with the correct name and value', () => {
        // The title must equal the value
        expect(hCardPropertyNode.prop('title')).toEqual(propertyValue);

        // The node must have the name set as a class
        expect(hCardPropertyNode.hasClass(propertyName)).toBeTruthy();

        // The value must also be set as the text
        expect(hCardPropertyNode.text()).toEqual(propertyValue);
    });
});

describe('once the hcard property with nesting has rendered', () => {
    let hCardPropertyNode;
    const propertyName = 'property-name';
    const nestedProperty = <HCardProperty name="nested-property" value="Nested Value" />;

    beforeEach(() => {
        hCardPropertyNode = shallow(
            <HCardProperty name={propertyName}>
                { nestedProperty }
            </HCardProperty>,
        );
    });

    it('renders with nested state correctly', () => {
        expect(hCardPropertyNode).toMatchSnapshot();
    });

    it('renders with the correct name and nested element', () => {
        // The node must have the name set as a class
        expect(hCardPropertyNode.hasClass(propertyName)).toBeTruthy();

        // The children must be rendered
        expect(hCardPropertyNode.containsMatchingElement(nestedProperty));
    });
});
