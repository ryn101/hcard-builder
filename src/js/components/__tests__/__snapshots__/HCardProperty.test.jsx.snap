// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once the hcard property has rendered renders with state correctly 1`] = `
<span
  className="property-name"
  title="Property Value"
>
  Property Value
</span>
`;

exports[`once the hcard property with nesting has rendered renders with nested state correctly 1`] = `
<span
  className="property-name"
  title=""
>
  <HCardProperty
    as="span"
    className=""
    name="nested-property"
    value="Nested Value"
  />
</span>
`;
